repo/fsrepo/migrations/versions.go

```diff
@@ -0,0 +1,93 @@
+package migrations
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"os/exec"
+	"path"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/coreos/go-semver/semver"
+)
+
+const distVersions = "versions"
+
+// LatestDistVersion returns the latest version, of the specified distribution,
+// that is available on the distribution site.
+func LatestDistVersion(ctx context.Context, fetcher Fetcher, dist string, stableOnly bool) (string, error) {
+	vs, err := DistVersions(ctx, fetcher, dist, false)
+	if err != nil {
+		return "", err
+	}
+
+	for i := len(vs) - 1; i >= 0; i-- {
+		ver := vs[i]
+		if stableOnly && strings.Contains(ver, "-rc") {
+			continue
+		}
+		if strings.Contains(ver, "-dev") {
+			continue
+		}
+		return ver, nil
+	}
+	return "", errors.New("could not find a non dev version")
+}
+
+// DistVersions returns all versions of the specified distribution, that are
+// available on the distriburion site.  List is in ascending order, unless
+// sortDesc is true.
+func DistVersions(ctx context.Context, fetcher Fetcher, dist string, sortDesc bool) ([]string, error) {
+	rc, err := fetcher.Fetch(ctx, path.Join(dist, distVersions))
+	if err != nil {
+		return nil, err
+	}
+	defer rc.Close()
+
+	prefix := "v"
+	var vers []*semver.Version
+
+	scan := bufio.NewScanner(rc)
+	for scan.Scan() {
+		ver, err := semver.NewVersion(strings.TrimLeft(scan.Text(), prefix))
+		if err != nil {
+			continue
+		}
+		vers = append(vers, ver)
+	}
+	if scan.Err() != nil {
+		return nil, fmt.Errorf("could not read versions: %s", scan.Err())
+	}
+
+	if sortDesc {
+		sort.Sort(sort.Reverse(semver.Versions(vers)))
+	} else {
+		sort.Sort(semver.Versions(vers))
+	}
+
+	out := make([]string, len(vers))
+	for i := range vers {
+		out[i] = prefix + vers[i].String()
+	}
+
+	return out, nil
+}
+
+// IpfsRepoVersion returns the repo version required by the ipfs daemon
+func IpfsRepoVersion(ctx context.Context) (int, error) {
+	out, err := exec.CommandContext(ctx, "ipfs", "version", "--repo").CombinedOutput()
```

Comment:
What is this used for, it doesn't seemed to be called anywhere?

Also, using exec like this is generally a bad idea since they user could have renamed their binary (e.g. ipfs-update will automatically at the version number to the end of binaries that are fetched). We could read the version corresponding to this package from the Go code, or if we wanted to read the repo version we could read it from the file itself (which would require passing the ipfs repo folder in too).

Reply:
I will move the exec portion back into ipfs-update to be used if IPFS_PATH is not set.

Reply:
Done.